Fix/27544 Incorrect Old Dot comment formatting when viewed on New Dot

Victor-Nyagudi opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

When viewing a report approved/submitted on Old Dot in New Dot, it was incorrectly formatted i.e. bolded, white, and separated on different lines. 

This PR restores the grey color and regular font weight similar to other types of report actions/system messages and shows the approved/submitted report on one line. 

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/27544
PROPOSAL: https://github.com/Expensify/App/issues/27544#issuecomment-1752142100

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

Pre-requisite: User A must have created a "Collect" workspace on Old Dot whose "Approval Mode" is set to "Submit and Approve", is the admin of the workspace, and has invited User B as an employee. 

1. As User B, click on "Reports" in the left sidebar and then click on the green pill button at the top right of the screen that says "New report" in the workspace User A invited you to.
2. Type a message in the composer at the bottom of the card that appears e.g. "Hello from the Old Dot workspace" and press Enter to send it.
3. In the browser Url, you'll see an object like `{"pageReportID":8096248714236998,"keepCollection":true}`. Copy the number after `pageReportID`.
4. Open a new browser window, log in as User A on New Dot, and paste the number you just copied from Old Dot in the browser Url to navigate to the report. It should look something like this: `staging.new.expensify.com/r/[the long number you copied goes here not surrounded by brackets]`.
5. Verify the message sent from Old Dot by User B is present and formatted correctly e.g. like a regular chat message - white in color and regular font weight. If you're not sure what this looks like, click on the large green button on the bottom left of the screen -> start chat -> enter an email or select from list -> type something in the composer and then send. 
6. Go back to the browser window where you're logged in as User B on Old Dot and add an expense to the report you created in step 1. Set User A as the person you're submitting to and ensure it's in the workspace you were invited to. 
7. Verify the text "You submitted this report to User A" is shown above the composer in the card at the centre of the screen right after your first message as User B. 
8. Go back to the New Dot window where you're logged in as User A viewing the report from Old Dot. You may need to refresh the page for the report to update. 
9. Verify the text "User B submitted this report to you" is shown after the first message User B sent and is greyed out, has a regular font weight, and is all on one line. 
10. Go back to Old Dot where you're logged in as User A and open the report User B created. 
11. Verify the chat history in the card is the same as the one you saw as User A in New Dot. 
12. Click on the green pill button at the top left that says "Approve". 
13. Verify the text "You final approved this report" appears just above the composer as the most recent message in the large card at the center of the screen. 
14. Go back to New Dot where you logged in as User A viewing the report of the workspace created in Old Dot. You might need to refresh the page. 
15. Verify the message "You final approved this report" is shown as the most recent message just above the composer and is greyed out, has a regular font weight, and is all in one line. 
16. Send a message in the report while on New Dot as User A e.g. "This is a message from New Dot".
17. Verify the message you just sent is formatted correctly like in step 5. 
18. If you don't see the conversation in the LHN in New Dot while on mobile devices or the desktop app, with the native app/browser open on mobile or the desktop app open, send a message in the workspace in New Dot from Mac Chrome.  You should see the conversation with the most recent message in the LHN. 
19. On Mac Chrome, Mac Safari, and Mac Desktop, resize the browser/app to different sizes. On Android Chrome and iOS Safari, switch between portrait mode and landscape mode. 
20. Verify all messages are formatted correctly as you've seen so thus far. 

**Please note:** 

You'll notice that the copy to clipboard icon appears when you hover over the system messages from Old Dot while in New Dot, however, clicking it doesn't copy anything to the clipboard. Try doing so and pasting what you "copied" in the composer. 

If in Android Web DEV (connected using your Mac's IP address and port number) you try long-pressing a message (in other chats too other than the workspace here), choose "Copy to clipboard" , and then try pasting the text in the composer, what you tried copying isn't pasted.

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
After repeating the steps in "Tests" above, turn off your wifi or go to your avatar in the top left -> preferences -> toggle force offline. 

Verify the text "You final approved this report" and "User B submitted this report to you" is greyed out, has a regular font weight, and is all in one line. 

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment. 

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
Same as above in Tests section. 

**Additional Steps (after performing steps in Tests section)**

1. While on New Dot, send a message with a link included in the workspace created on Old Dot e.g. "Hello www.google.com".
2. Verify the link is blue and underlined. 
3. Send a message in the same workspace formatted to use markdown e.g. bold, italic, strikethrough, a mix of markdown styles, etc. 
4. Verify the appropriate markdown style is applied. 
5. Hover over the messages you sent in Step 1 and 3, click on the pencil icon, and edit the message. Press "Enter" when done. 
6. Verify the edited text is what you edited the message to, and the small "edited" label is applied. 
7. Hover over the "You final approved this report" and "User B submitted this report to you" messages and click on one of the emojis. 
8. Verify the emojis are shown below the messages. 
9. Copy the link in the Url, sign out of New Dot, paste it in the browser, and then press Enter. 
10. Log into your account. 
11. Verify you're directed to the Old Dot workspace page, and everything looks the way you left it, including the "You final approved this report" and "User B submitted this report to you" messages being greyed out, having a regular font weight, and appearing on one line. 

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->
![expensify-android-native-fixed-OD-message-font.jpg](https://github.com/Expensify/App/assets/79470910/e3c92afe-f343-4ba9-965e-844b0f6f3282)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->
![expensify-OD-comment-fix-android-web.jpg](https://github.com/Expensify/App/assets/79470910/ba20d545-0f6f-4e2d-aab9-eb3689cfdc7f)

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->
![expensify-OD-comment-PR-fix-iOS-Native](https://github.com/Expensify/App/assets/79470910/b5f295c4-4f63-4c67-a363-05ac4000e4f6)

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->
![expensify-OD-comment-PR-fix-iOS-Web](https://github.com/Expensify/App/assets/79470910/c9bec0f6-7d0e-4e15-86e7-cf4d2ee159ac)

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->
![expensify-OD-comment-PR-fix-Mac-Safari](https://github.com/Expensify/App/assets/79470910/ec272818-dbdf-471b-9a65-d28c9205095c)

![expensify-OD-comment-PR-fix-Mac-Chrome](https://github.com/Expensify/App/assets/79470910/4fc299cb-2330-4f57-a5b6-278f2c51d919)

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->
![expensify-OD-comment-PR-fix-Mac-Desktop](https://github.com/Expensify/App/assets/79470910/e10c8483-f00a-41c1-842b-bc67d8d55a84)

</details>

Victor-Nyagudi committed:
Wrap ReportActionItemFragment inside a Text component

Victor-Nyagudi committed:
Restore grey color and regular font weight to APPROVED/SUBMITTED messages

Victor-Nyagudi committed:
Conditionally surround ReportActionItemFragment with Text component

Victor-Nyagudi committed:
Remove return statement surrounding ReportActionItemFragment

This was causing a lint error during PR checks

Victor-Nyagudi committed:
Surround 'props.fragment.text' in curly braces

Caught this little error currently on latest main that would've rendered 'props.framgent.text' literally instead of the actual text

Victor-Nyagudi commented:
~Caught this little bug currently on the latest main that would've literally rendered 'props.fragment.text' instead of the actual text.~ 
Found this piece of code that looked a bit suspicious on latest main. With the current implementation, it would literally render 'props.fragment.text`

https://github.com/Expensify/App/blob/d28247332f58bb0d4296b0eab49527771c03464c/src/pages/home/report/ReportActionItemFragment.js#L185

~I've fixed it like so in the PR.~ 
I've changed it to the following in my PR.

```javascript
<Text>{props.fragment.text}</Text>
```

Seeing how the other cases are handled, it's possible this may have been intentional. 

https://github.com/Expensify/App/blob/60576fe3d3a4c91cc33192c01ef213b59a336441/src/pages/home/report/ReportActionItemFragment.js#L170-L183

If at all any of you, @jjcoffee or @lakchote, feel this change isn't needed, I can just revert it. It could also save on some testing.

Victor-Nyagudi commented:
Just Android Native testing left to go, and the PR should be ready for review. Saved it for last because [it can take a while to build](https://expensify.slack.com/archives/C01GTK53T8Q/p1696657978715369?thread_ts=1696601366.090489&cid=C01GTK53T8Q).

Victor-Nyagudi committed:
Describe function using JS Docs syntax & use inline ternary statement

Victor-Nyagudi committed:
Revert "Surround 'props.fragment.text' in curly braces"

This reverts commit 9c0677861ff6b7eddd86ad0fe3d4ba8a61351944.

Victor-Nyagudi commented:
Tried building from Android Studio, but that took a long time to even import the files, so I switched back to using VS Code. 

Still waiting on the app to build on Android Native.

Victor-Nyagudi commented:
Android Native has built, but the font is still bolded, for some reason. The color is grey as expected.

Investigating.

Victor-Nyagudi commented:
It seems that's just how a`fontWeight: 'normal'` looks on Android because when I tried changing it to `fontWeight: 'bold'`, the text got bigger.

Victor-Nyagudi commented:
The [styling guide](https://github.com/Expensify/App/blob/main/contributingGuides/STYLING.md#inline-styles) says inline styles are forbidden, but I was a bit hesitant to create a helper method in `styles.ts` to add just two styles `ReportActionItemFragment`, so I figured I'd ask first.

An inline style has been applied over here, so I thought maybe things could work as is. 

https://github.com/Expensify/App/blob/3380266055d37f7cc4521cf31649170597fc5fc0/src/components/MentionSuggestions.js#L107

In any case, should we go with the helper method, I could add this in the `styles` variable of `styles.ts`.

```javascript
getMutedColorAndNormalFontWeight: (isApprovedOrSubmittedReportActionType: Boolean) => (isApprovedOrSubmittedReportActionType ? {color: theme.textSupporting, fontWeight: 'normal'} : {}),
```

And then replace the added style in `ReportActionItemFragment` with this:

```javascript
style={[
    styles.chatItemMessageHeaderSender,
    props.isSingleLine ? styles.pre : styles.preWrap,
    styles.getMutedColorAndNormalFontWeight(_.contains([CONST.REPORT.ACTIONS.TYPE.APPROVED, CONST.REPORT.ACTIONS.TYPE.SUBMITTED], props.actionName)),
]}
```

Victor-Nyagudi committed:
Merge branch 'Expensify:main' into fix/27544-incorrect-OD-comment-formatting

jjcoffee commented:
>If you hover anywhere over the messages "You final approved this report" or "User B submitted this report to you" in the Old Dot workspace while on New Dot, a tooltip with no Avatar is displayed.

@Victor-Nyagudi Regarding the above, to keep things consistent with other system messages, I actually think we don't want tooltips to display here. Is it an easy fix? I think technically it's out of scope for this issue, but it might as well be fixed if it's not too big a deal!

Victor-Nyagudi commented:
@jjcoffee The fragments we get from Old Dot are wrapped in `UserDetailsTooltip`. 

I suspect this could be why the tooltip is shown on hover, though I haven't dived too deep into it.

https://github.com/Expensify/App/blob/e6af6e9df42b1394747cdd30c31085e65bdaa243/src/pages/home/report/ReportActionItemFragment.js#L155-L169

It's EOD for me (I can still reply on mobile), so I'll take a deeper look first thing tomorrow to try figure out why the tooltip is there in the first place and what we can potentially do about it.

jjcoffee commented on src/pages/home/report/ReportActionItemFragment.js:
```diff
@@ -63,6 +64,9 @@ const propTypes = {
     isThreadParentMessage: PropTypes.bool,
 
+    /** The report's action name/type e.g. APPROVED, SUBMITTED, etc. */
+    actionName: PropTypes.string,
```
I think we can just probably pass `isApprovedOrSubmittedReportActionType` in `ReportActionItemFragment` so that we don't have to recheck it here.

Victor-Nyagudi replied:
Made the change here 1d698f3d0a362bd67fd46ee48fcbd37c14a8004f.

Victor-Nyagudi replied:
https://github.com/Expensify/App/pull/30193/files/64d204b196a8295ad037b97a967702cea6245d17#diff-7451957e0135ee20216b214d4e27202271aeacbd2e321b9b35b5b1e60d6677b6

jjcoffee commented on src/pages/home/report/ReportActionItemMessage.js:
```diff
@@ -47,25 +48,42 @@ function ReportActionItemMessage(props) {
-                    />
-                ))
+            {isApprovedOrSubmittedReportActionType ? (
+                // Wrapping 'ReportActionItemFragment' inside '<Text>' so that text isn't broken up into separate lines when
+                // there are multiple messages of type 'TEXT', as seen when a report is submitted/approved from a
+                // policy on Old Dot and then viewed on New Dot.
```
```suggestion
                // Approving or submitting reports in oldDot results in system messages made up of multiple fragments of `TEXT` type
                // which we need to wrap in `<Text>` to prevent them rendering on separate lines.
```

Victor-Nyagudi committed:
Reword comment explaining reason for wrapping ReportActionItemFragment with <Text>

This is done as per @jjcoffee's suggestion.

Co-authored-by: Joel Davies <joeld.dev@gmail.com>

Victor-Nyagudi committed:
Merge branch 'Expensify:main' into fix/27544-incorrect-OD-comment-formatting

Victor-Nyagudi committed:
Pass isApprovedOrSubmittedReportActionType prop instead of actionName to ReportActionItemFragment

Victor-Nyagudi committed:
Remove unused import statement

Victor-Nyagudi commented:
> Regarding the above, to keep things consistent with other system messages, I actually think we don't want tooltips to display here. Is it an easy fix? I think technically it's out of scope for this issue, but it might as well be fixed if it's not too big a deal!

@jjcoffee The tooltip was introduced to [`ReportActionItemFragment`](https://github.com/Expensify/App/commit/cf86c6f6b2c00415857da8daf832bc94ad3e2d7b#diff-7451957e0135ee20216b214d4e27202271aeacbd2e321b9b35b5b1e60d6677b6) on [March 4th, 2021](https://github.com/Expensify/App/commit/cf86c6f6b2c00415857da8daf832bc94ad3e2d7b) in this PR by @parasharrajat.

- https://github.com/Expensify/App/pull/1632

The Expensify app looked very different back then, as you can see from the video in the attached PR, and a lot has happened since. 

It's also possible the databases weren't synced or were separate at the time, so reports from Old Dot weren't making it to New Dot. 

That being said, the tooltip appearing on messages from Old Dot could be a regression from the mentioned PR, unless maybe @parasharrajat could provide us with some more context on why this change was made. 

In any case, I came up with two solutions - one a bit riskier than the other. 

- Option 1 (riskier)

Remove the `UserDetailsTooltip` wrapper so just the text is rendered. 

```javascript
// ...
case 'TEXT':
    return (
        <Text
            numberOfLines={props.isSingleLine ? 1 : undefined}
            style={[
                styles.chatItemMessageHeaderSender,
                props.isSingleLine ? styles.pre : styles.preWrap,
                props.isApprovedOrSubmittedReportActionType && {color: styles.colorMuted.color, fontWeight: 'normal'},
            ]}
        >
             {props.fragment.text}
        </Text>
);
// ...
```

- Option 2

Conditionally render the tooltip based on whether the report action's type is `APPROVED` or `SUBMITTED`.

We'll do so by first passing a `isApprovedOrSubmittedReportActionType` prop to `UserDetailsTooltip`.

```javascript
// ...
<UserDetailsTooltip
    accountID={props.accountID}
    delegateAccountID={props.delegateAccountID}
    icon={props.actorIcon}
    isApprovedOrSubmittedReportActionType={props.isApprovedOrSubmittedReportActionType} // <- Added
>
//...
```

We'll then use this prop in [`BaseUserDetailsTooltip.web.js`](https://github.com/Expensify/App/blob/414d5913999f69d83ab08b6ae6e03325c9ede549/src/components/UserDetailsTooltip/BaseUserDetailsTooltip.web.js#L65-L74) to decide whether or not to surround `UserDetailsTooltip`'s children with `Toolitp`.

```javascript
// ...
return props.isApprovedOrSubmittedReportActionType ? (
    props.children
) : (
    <Tooltip
        shiftHorizontal={props.shiftHorizontal}
        renderTooltipContent={renderTooltipContent}
        renderTooltipContentKey={[userDisplayName, userLogin]}
        shouldHandleScroll
    >
        {props.children}
    </Tooltip>
);
// ...
``` 

<details>

<summary>Short demo after implementing option 2</summary>

https://github.com/Expensify/App/assets/79470910/f0d2c565-4ca4-4fc5-a20f-2e5a563b880b

</details>

Victor-Nyagudi commented:
Hi, @parasharrajat 

When you approve or submit a report in a workspace in Old Dot and then view the report in New Dot, a tooltip with no avatar is shown when hovering over the message.

The message is also broken up into fragments, so a tooltip is shown when hovering over each fragment.

We were curious why reports of `case: TEXT` in `ReportActionItemFragment` were wrapped in `UserDetailsTooltip` in the first place.

jjcoffee commented:
@Victor-Nyagudi Usernames are displayed as a `TEXT` type fragment, that's why they're wrapped in the `UserDetailsTooltip`. I think we just want to conditionally wrap for anything that's not `isApprovedOrSubmittedReportActionType`.

jjcoffee commented:
@Victor-Nyagudi Whilst I appreciate the level of detail in your testing steps, I think it's a bit overkill. Can you try and reduce the number of steps so that it's closer to the original [issue](https://github.com/Expensify/App/issues/27544)? The QA team will know how to do basic things like creating expense reports in oldDot, so it's not necessary to explain in so much detail.

Victor-Nyagudi commented:
> Whilst I appreciate the level of detail in your testing steps, I think it's a bit overkill. Can you try and reduce the number of steps so that it's closer to the original https://github.com/Expensify/App/issues/27544?

@jjcoffee I've reduced the number of steps as much as possible without losing any clarity. 

The original issue may have had fewer steps, but it wasn't very clear. It doesn't state anywhere that User B needs to submit a report while on Old Dot instead of New Dot. I only knew this was the case when someone asked for clarity in this [Slack thread you created](https://expensify.slack.com/archives/C01GTK53T8Q/p1696602918102179?thread_ts=1696601064.364559&cid=C01GTK53T8Q).

The issue also says to change the workspace approval mode to "Submit and close" while in order to _approve_ a message, the approval mode needs to be "Submit and approve".

I wasn't the only one having trouble reproducing this issue, as you can see from the Slack thread, so I decided not to leave anything to chance with the QA team. 

I hope this works.

Victor-Nyagudi committed:
Wrap initial ReportActionItemFragments in Text

This commit also gets rid of the renderReportActionItemFragment() method and flaggedContentText variable that are no longer necessary

Victor-Nyagudi commented on src/pages/home/report/ReportActionItemMessage.js:
```diff
@@ -47,25 +48,41 @@ function ReportActionItemMessage(props) {
-                        displayAsGroup={props.displayAsGroup}
-                    />
-                ))
+            {isApprovedOrSubmittedReportActionType ? (
```
Good catch!

I also felt a little uneasy about repeating the inline conditional. 

This commit addresses your concerns.  20e26f0f912eb73fc207080d6a6b934429132961.

I've gone back to using `!props.isHidden` conditional and the flagged content text is no longer double wrapped. The variable itself actually isn't needed anymore along with the `renderReportActionItemFragment()` method.

Victor-Nyagudi committed:
Ran prettier again so linter is happy

Victor-Nyagudi committed:
Create helper method to replace inline stlyes

Victor-Nyagudi committed:
Rename isApprovedOrSubmittedReportActionType

The 'type' at the end felt redundant. This can inferred from reading the comments above the prop types wherever it's used

Victor-Nyagudi commented:
I'm getting conflicting approaches to handling the tooltip issue. 

@jjcoffee says:

> I don't think it makes much sense to pass a flag to UserDetailsTooltip when we can just not wrap with it.

@parasharrajat says: 

> Now, that we are rendering more things with this fragment, let's move the tooltips to another place where are only rendering the actor name.

Which approach should I go with?

I also don't get you quite clearly, @parasharrajat. Where exactly should we move the tooltip to? Who is the actor?

Victor-Nyagudi committed:
Merge branch 'Expensify:main' into fix/27544-incorrect-OD-comment-formatting

Victor-Nyagudi commented:
@jjcoffee With regard to your [comment](https://github.com/Expensify/App/pull/30193#discussion_r1373060857) on using `!props.isHIdden` conditional in `ReportActionItemMessage`, we could, in addition, also use `isApprovedOrSubmittedReportAction` to conditionally wrap the fragments in `Text`.

We'll need to reintroduce the `renderReportActionItemFragment()` method and store the fragments in a variable. Here's what the code will look like after the changes. 

```javascript
// ...
/**
    * Get a ReportActionItemFragment
    * @param {Object} fragment the current message fragment
    * @param {Number} index the current message fragment's index
    * @returns {Object} report action item fragment
*/
const renderReportActionItemFragment = (fragment, index) => (
    <ReportActionItemFragment
        key={`actionFragment-${props.action.reportActionID}-${index}`}
        fragment={fragment}
        iouMessage={iouMessage}
        isThreadParentMessage={ReportActionsUtils.isThreadParentMessage(props.action, props.reportID)}
        attachmentInfo={props.action.attachmentInfo}
        pendingAction={props.action.pendingAction}
        source={lodashGet(props.action, 'originalMessage.source')}
        accountID={props.action.actorAccountID}
        style={props.style}
        displayAsGroup={props.displayAsGroup}
        isApprovedOrSubmittedReportAction={isApprovedOrSubmittedReportAction}
    />
);

const reportActionItemFragments = _.map(messages, (fragment, index) => renderReportActionItemFragment(fragment, index));

return (
    <View style={[styles.chatItemMessage, !props.displayAsGroup && isAttachment ? styles.mt2 : {}, ...props.style]}>
        {!props.isHidden ? (
            // Approving or submitting reports in oldDot results in system messages made up of multiple fragments of `TEXT` type
            // which we need to wrap in `<Text>` to prevent them rendering on separate lines.

            isApprovedOrSubmittedReportAction ? (
                <Text>{reportActionItemFragments}</Text>
            ) : (
                 <>{reportActionItemFragments}</>
            )
       ) : (
           <Text style={[styles.textLabelSupporting, styles.lh20]}>{props.translate('moderation.flaggedContent')}</Text>
       )}
    </View>
);
// ...
```

Let me know if these changes work for you.

Victor-Nyagudi commented:
> I think we just want to conditionally wrap for anything that's not `isApprovedOrSubmittedReportActionType`.

@jjcoffee This is what the code could look like with this approach. 

```javascript
// ...
case 'TEXT': {
    const messageFragment = (
        <Text
            numberOfLines={props.isSingleLine ? 1 : undefined}
            style={[styles.chatItemMessageHeaderSender, props.isSingleLine ? styles.pre : styles.preWrap, styles.approvedOrSubmittedMessage(props.isApprovedOrSubmittedReportAction)]}
        >
            {props.fragment.text}
        </Text>
    );

    return props.isApprovedOrSubmittedReportAction ? (
        messageFragment
    ) : (
        <UserDetailsTooltip
            accountID={props.accountID}
            delegateAccountID={props.delegateAccountID}
            icon={props.actorIcon}
        >
            {messageFragment}
       </UserDetailsTooltip>
    );
}
// ...
```

This also tests well. 

Let me know if these changes work for you so we can keep this PR moving forward.

Victor-Nyagudi committed:
Move styling helper method to StyleUtils

This felt like the most logical place to put the helper method seeing how all the other helper methods are located here.

Victor-Nyagudi committed:
Conditionally wrap fragment text in a tooltip

Victor-Nyagudi commented:
I wasn't quite satisfied with how things looked on Android Native, so I did some more digging to see if I could make some improvements.

I discovered the `chatItemMessageHeader` style gives the fragment text a `fontFamily: font Family.EXP_NEUE_BOLD`.

This is what caused the text to still appear in bold on Android.

I changed it to `fontFamily: fontFamily.EXP_NEUE` in this commit, , in addition to the font weight change for other platforms, and it tested well.

Here's how it looks on Android Native now.

![expensify-android-native-fixed-OD-message-font.jpg](https://github.com/Expensify/App/assets/79470910/78f066c0-0aef-4c07-870b-6dcf4d349fbd)

I've also updated the screenshot in the PR checklist to reflect these changes.

Victor-Nyagudi commented:
@jjcoffee  With the changes I've made so far, the following should be addressed by now:

- Reports from Old Dot not formatted properly
- Tooltip appearing on Old Dot fragments
- Text still appearing in bold on Android native

I think we're in a good spot right now, and all the tests still pass. 

The only thing I see left is your opinion on this [comment](https://github.com/Expensify/App/pull/30193#issuecomment-1783737398), but otherwise, I think we should be good to go.

Victor-Nyagudi committed:
Merge branch 'main' into fix/27544-incorrect-OD-comment-formatting

Victor-Nyagudi committed:
Run pretteir to fix diffs

Victor-Nyagudi committed:
Merge branch 'main' into fix/27544-incorrect-OD-comment-formatting

Victor-Nyagudi committed:
Remove check in getApprovedOrSubmittedReportTextStyles

Since this helper method is only used in one place so far, it's better to conditionally call it in the place it's being used rather than have the check in the method and return an empty object

Victor-Nyagudi committed:
Merge branch 'Expensify:main' into fix/27544-incorrect-OD-comment-formatting

jjcoffee commented on src/pages/home/report/ReportActionItemFragment.js:
```diff
@@ -152,21 +157,32 @@ function ReportActionItemFragment(props) {
+                <Text
+                    numberOfLines={props.isSingleLine ? 1 : undefined}
+                    style={[
+                        styles.chatItemMessageHeaderSender,
```
I don't think we really want to reuse the same `<Text>` component and props here as  `styles.chatItemMessageHeaderSender` isn't appropriate for the system message. I also think we need to have some handling for RTL as in the `COMMENT` case:

https://github.com/Expensify/App/blob/cf006036f4e8e480027b9e5c0b9b847bfd41f10d/src/pages/home/report/ReportActionItemFragment.js#L127-L138

Otherwise we end up with this issue:

![image](https://github.com/Expensify/App/assets/27287420/78dae9ae-ee60-4b47-9edd-8b8ef72b7e4a)

Victor-Nyagudi replied:
Some of the styles in `styles.chatItemMessageHeaderSender` apply to the system message like the `fontSize`, `lineHeight`, etc. 

Here's how the text "marked as complete" in a task is styled when a task is marked complete. 

https://github.com/Expensify/App/blob/e1083781e4b98fa4c2c9cf49cf1742b7d89fa2ba/src/components/ReportActionItem/TaskAction.js#L24

Should we follow the same approach for the `<Text>` component that will be meant for system messages now that you prefer having two separate ones?

Are we also going to not use `props.isSingleLine` for said component?

jjcoffee replied:
My point is that you needed to add an extra style to override those in `styles.chatItemMessageHeaderSender`, which doesn't really make sense when we can just no use it here (since it's not a header). Seems like we could take inspiration from `TaskAction`, yes, and also for the `COMMENT` fragments (to handle RTL).

Victor-Nyagudi replied:
@jjcoffee This is what I'm getting so far.  

> I don't think we really want to reuse the same <Text> component

1. Since we're not reusing the component, there'll be two `Text` components - one for system messages and another that'll be wrapped in a tooltip as usual. The `style` prop for the system message will be similar to the one in `TaskAction` so that we're not overwriting `styles.chatItemMessageHeaderSender`, but the remaining props will be similar to the `Text` wrapped in a tooltip.

2. The `Text` component for system messages needs to handle RTL like regular messages do (I'd also like to know how you got the text to start from right to left for testing).

3. Since the system message will have its own `Text` component styled using existing styles like in `TaskAction`, the helper method will no longer be necessary and will be removed.

Does all this sound right 👍?

jjcoffee replied:
Sounds good! You can get RTL by entering e.g. an arabic display name.

Victor-Nyagudi replied:
@jjcoffee Does this look right to you for an Arabic display name?

![expensify-OD-system-message-with-arabic-name](https://github.com/Expensify/App/assets/79470910/03ba0476-db2c-4bfb-bc4c-9da93dd46c9d)

jjcoffee replied:
I think the bracket is meant to come after

Victor-Nyagudi replied:
How about now?

![expensify-OD-correct-rtl-format](https://github.com/Expensify/App/assets/79470910/9483bc10-aa99-45c4-8266-4cf610181424)

jjcoffee replied:
Yup looks good! Btw you can verify on other system messages, e.g. inviting a user with an Arabic display name to a workspace.

Victor-Nyagudi replied:
Good to know. Still testing on other platforms.

Victor-Nyagudi replied:
Just iOS left to test with my new solution. 

Ran into a few hiccups, but I'm back on the right track.

Victor-Nyagudi replied:
I'm grateful you had [concerns](https://github.com/Expensify/App/pull/30193#issuecomment-1790397963) about indiscriminately wrapping the fragments in `Text` because I discovered this would've caused a regression on Android native where really long messages overflowed the screen's bounds. 

This wasn't happening earlier, or maybe I just didn't notice, but either way, I found a way to conditionally wrap the fragments in `Text` so we have less to worry about in terms of regressions. 

All this is done in this commit f669405e5edda8668d0c4dbfd1d2e56c37622f91.

I also used separate `Text` components in `ReportActionItemFragment` so we're not overwriting styles, added styling to account for RTL scenarios, and got rid of the helper method in `StyleUtils` since it's no longer needed. 

Everything tests well, and I hope it's the same on your end. 

Hopefully, it's not EOD for you yet, and we can still hit the soft [deadline](https://github.com/Expensify/App/issues/27544#issuecomment-1790406897) you set earlier, but if things still feel amiss, feel free to let me know and we can tighten things up even further.

jjcoffee commented:
@Victor-Nyagudi Sorry for the delay, I got swamped! In response to your [question](https://github.com/Expensify/App/pull/30193#issuecomment-1783737398), it sounds safest to add the `isApprovedOrSubmittedReportAction` conditional so let's go with that.

Victor-Nyagudi commented:
@jjcoffee It appears the linter doesn't like nested ternary expressions as I [suggested](https://github.com/Expensify/App/pull/30193#issuecomment-1783737398), but the [docs](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#inline-ternaries) also advise against storing the result of a ternary operation in a variable. 

I'm open to suggestions on how we can implement the optional wrapping in `Text` using `isApprovedOrSubmittedReportAction`, but I'm also curious why you feel the current implementation isn't as safe. 

Since `ReportActionItemFragment` always renders a `Text` component, it should be valid React syntax, and the child `Text` components will still align for the OD system messages.

Victor-Nyagudi commented:
Also, let me know what times/if you're available to collaborate in real-time on this PR so we can get it across the finish line.

I think we're close.

jjcoffee commented:
> I'm open to suggestions on how we can implement the optional wrapping

Hmm did you consider just adding flex-direction: row conditionally on the parent `View`?

https://github.com/Expensify/App/blob/3db5ff4a1924156c748291d25eb9d774d82f879e/src/pages/home/report/ReportActionItemMessage.js#L51

I think it's just a bit messy to unnecessarily wrap with a `<Text>` when we don't need to, but if there's no other way we can stick to it, I guess!

Victor-Nyagudi commented:
Yes, I did explore styling the `View` with `flexDirection: 'row'` in my [proposal](https://github.com/Expensify/App/issues/27544#issuecomment-1752142100), but this caused issues when the screen was  resized.

![image](https://github.com/Expensify/App/assets/79470910/655096fc-d54f-479f-9f81-bcf4a0d3d5ca)

jjcoffee commented:
@Victor-Nyagudi Ah sorry, forgot about that! What about just `display: inline` then? The [RTL](https://github.com/Expensify/App/pull/30193/files#r1378636848) issue seems to be present when using a `<Text>` wrapper (even if setting LTR styles on the children).

Victor-Nyagudi commented:
No worries. 

Did you notice any issues when the fragments aren't wrapped in a `<Text>` component? 

I'm investigating as we speak.

Victor-Nyagudi committed:
Merge branch 'Expensify:main' into fix/27544-incorrect-OD-comment-formatting

Victor-Nyagudi committed:
Conditionally wrap report action fragments in Text and account for RTL scenarios

Arabic display names caused the text to start from the right side of the screen even though the system message is in English

Victor-Nyagudi commented:
Looks like one of the checks failed. 

I'll investigate first thing in the morning.

Victor-Nyagudi commented:
It appears the performance test started failing after I merged `main` here 8afe93c58094d548e2639276d723e961a80d789a before my most recent changes.

These tests also seem to be [fairly new](https://expensify.slack.com/archives/C01GTK53T8Q/p1698958071295039?thread_ts=1698952935.068319&cid=C01GTK53T8Q). Could we try running them again to see if they're still failing?

@jjcoffee We're also approaching the 9 business days mark, so this PR is now my top priority so we can get it merged before that point is reached.

jjcoffee commented on src/pages/home/report/ReportActionItemFragment.js:
```diff
@@ -63,6 +63,12 @@ const propTypes = {
+    isApprovedOrSubmittedReportAction: PropTypes.bool,
+
+    /** Used to format RTL display names in Old Dot system messages e.g. Arabic */
+    shouldConvertToLTR: PropTypes.bool,
```
I'm not sure why we need this? I think we always want to convert to LTR!

jjcoffee replied:
Hmm strange I'm not getting the same

Victor-Nyagudi replied:
Interesting. 

Are you getting the correct result or the wrong one?

jjcoffee replied:
Ah sorry for not being clearer, I meant I don't see the same issue if every fragment is converted to LTR. Can you double check it's necessary to add the `shouldConvertToLTR`?

jjcoffee replied:
It feels a bit messy using `shouldConvertToLTR`, as really we're passing `isFirstFragment`. If any other fragment wanted to use `shouldConvertToLTR` it might be a bit unexpected that it's entirely based on the index :smile: So if it is necessary I suggest either renaming or maybe just passing the fragment index instead?

jjcoffee replied:
No need to spend super long on it though, if you're still getting the same result with converting to LTR on all fragments maybe it's some specific length of text that messes it up for some reason :smile:

Victor-Nyagudi replied:
> Can you double check it's necessary to add the shouldConvertToLTR?

@jjcoffee  I've just tested on a physical device, and yes, the check is necessary for the text to be properly formatted. 

<details>

<summary>Latest screenshot after double checking</summary>

</details>

> So if it is necessary I suggest either renaming or maybe just passing the fragment index instead?

I'll go ahead and rename the prop to `isFragmentContainingDisplayName` which is a bit more descriptive, if that's okay.

jjcoffee replied:
Huh interesting! The trouble is `isFragmentContainingDisplayName` would return true for all first fragments, so could result in an unexpected result too. `isFirstFragment` is probably the best prop name as it fully describes what's in there :smile: In a way I think I prefer passing the index and making the check in the fragment itself, but don't feel super strongly so I'll leave it up to you.

Victor-Nyagudi replied:
I understand. 

The reason I went with `isFragmentContainingDisplayName` is it explains what fragment we're looking for.  

I was hesitant on using `isFirstFragment` because it could lead to the question "Why is the first fragment so important?" 

If at all something changes in the future and the first fragment is no longer the one containing the display name, `isFirstFragment` wouldn't make much sense, but `isFragmentContainingDisplayName` felt like it would hold pretty well regardless of which position the fragment containing the display name would be, and we'd just need to check the correct index.

jjcoffee commented on src/pages/home/report/ReportActionItemFragment.js:
```diff
@@ -152,8 +160,15 @@ function ReportActionItemFragment(props) {
+            return props.isApprovedOrSubmittedReportAction ? (
+                <Text
+                    numberOfLines={props.isSingleLine ? 1 : undefined}
+                    style={[styles.chatItemMessage, styles.colorMuted]}
```
I think we want `styles.ltr` here too, like for the `COMMENT` case above.

jjcoffee replied:
Oops reading things out of order, ignore me! :sweat_smile:

Victor-Nyagudi replied:
For what it's worth, adding `styles.ltr` there made no difference. 

Applying it to the wrapper `Text` in `ReportActionItemMessage` was enough.

jjcoffee commented:
@Victor-Nyagudi Just a heads up that it's about 3 hours until my EOD.

Victor-Nyagudi commented:
Got it, @jjcoffee . I'm available for all 3.

jjcoffee commented:
@Victor-Nyagudi Any luck with [this](https://github.com/Expensify/App/pull/30193#discussion_r1381553684)?

Victor-Nyagudi commented:
Checkign right now.

Victor-Nyagudi commented:
@jjcoffee Aside from the `shouldConvertToLTR` bit, is there anything else that requires my attention?

jjcoffee commented:
@Victor-Nyagudi Nope happy with everything else!

Victor-Nyagudi committed:
Rename shouldConvertToLTR prop to isFragmentContainingDisplayName

jjcoffee reviewed (approved):
LGTM!

Victor-Nyagudi commented:
Thank you for your patience and sticking with me through this one, @jjcoffee.

lakchote commented:
Great work and collaboration both of you @Victor-Nyagudi @jjcoffee  👍

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6777007434) to staging by https://github.com/lakchote in version: 1.3.96-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6818231417) to production by https://github.com/puneetlath in version: 1.3.96-15 🚀

platform | result
---|---
🤖 android 🤖|failure ❌
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6826733490) to production by https://github.com/puneetlath in version: 1.3.97-7 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
